c/src/decode.c

```
@@ -10,11 +10,36 @@
  * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
  */
 
+#include <string.h>
+
 #include <ubx/decode.h>
 
-#include <stdio.h>
 #include <swiftnav/bits.h>
 
+// UBX protocol is little-endian. This file assumes that the host system is also
+// little endian.
+// TODO(yizhe) write endian-independent code
+
+/** Writes checksum over `length` bytes of `buf` into `CK_A` and `CK_B`. The
+ * `length` includes class id, msg id, length bytes, and payload. Uses the 8-Bit
+ * Fletcher Algorithm.
+ *
+ * \param buff Buffer containing a ubx message.
+ * \param length Length of the ubx message
+ * \param CK_A first byte of checksum
+ * \param CK_B second byte of checksum
+ */
+void ubx_checksum(const uint8_t buff[], size_t length, uint8_t *checksum) {
+  uint8_t *CK_A = checksum;
+  uint8_t *CK_B = checksum + 1;
+  *CK_A = 0;
+  *CK_B = 0;
+  for (size_t i = 0; i < length; i++) {
+    *CK_A = *CK_A + buff[i];
```

This sorta kinda looks like a fletcher checksum? If that's what it's supposed to be I don't think this the correct implementation. A true fletcher checksum uses 1's complement math, so there are edge cases that need to be handled when running on hardware doing 2's complement math.

If this is actually how the UBX message checksum is calculated and it's not actually a fletcher checksum that's cool, they wouldn't be the first people to screw it up.

---

I guess to check this out I think it would be good to run as much data through this as possible and flag any messages with a checksum mismatch and then run those messages through a proper fletcher checksum implementation to see if they pass.

Messages with lots of data will be the most likely to catch edge cases, especially if we want get messages with lots of large value bytes in it. The edge case I'm worried about here is when `CK_A + buff[i] > 0x100` you end up losing a bit in the rollover.

---

Yep, this is supposed to be a fletcher checksum. 
This code is lifted directly out of the UBX protocol spec, with the specific callout to use `uint8_t` (i.e., they are relying on integer overflow)
If there's any delta between this and a proper fletcher, I expect the UBX protocol to follow this "incorrect" pattern

---

I think we should be verifying this. It looks like we're currently silently dropping packets that fail the checksum check, so we can't easily tell if we're just dropping packets because of a bad checksum algorithm or some other legit reason.

---

Verified, this version matches with UBX's checksums